### ISSUE TITLE
Stability issues

### ISSUE THREAD

**Issue by flyaruu** [sammhicks/picoserve#9]
I'm experiencing some issues when there are multiple concurrent requests.

https://github.com/flyaruu/esp32-http-api

This hello-world repo works fine on a esp32 as long as I keep the request speed reasonable, it keeps working, but if I go a bit wild (e.g. holding the reload button in a browser) it will stop accepting connections and never recover. 
Things that seem to help a little bit:
- More slots in the StackResource
- Returning a Connection: Close header

That the performance is limited, that is reasonable, but the breaking and never recovering is a bit of an issue.
Any ideas? Do you see it on other platforms as well? Or is it a esp32 thing?

**Comment by sammhicks** [comment]
I've just ran the hello_world_embassy example on a Raspberry Pi Pico W, setting "WEB_TASK_POOL_SIZE" to be 1, and thus StackResources to be 1, and hammered the connection by running

for i in $(seq 1 1000); do curl 192.168.0.1; echo done; done

Occasionally curl fails to connect, but it recovers afterwards. If I run the script in two terminals at the same time, it frequently fails to connect, but recovers afterwards. I suspect the problem might be an esp32 thing :(

**Comment by Token-Thinker** [comment]
Interested to see the outcome of the results (also awesome videos @Flyruu btw).

I'm working with the ESP32-Cam and I see similar issues when initializing the connection. It hangs if connection is missed or drops resulting in having to reset the device.

**Comment by sammhicks** [comment]
@flyaruu As a test, are you able to create a wrapper struct around `embassy_net::tcp::TcpWriter` which implements `embedded_io_async::Write`, with timeouts on write and flush? Then once you split the `embassy_net::tcp::TcpSocket`, wrap the write half with the new wrapper struct and pass that to `picoserve::serve_with_state`

**Comment by flyaruu** [comment]
Yes, that did the trick! I've added this wrapper:
```rust
pub struct WrappedWriter<'a,T: picoserve::Timer<Duration = embassy_time::Duration>>(pub TcpWriter<'a>, pub T);

impl <'a,T: picoserve::Timer<Duration = embassy_time::Duration>>ErrorType for WrappedWriter<'a, T> {
    type Error = embassy_net::tcp::Error;
}

impl <'a, T: picoserve::Timer<Duration = embassy_time::Duration>>Write for WrappedWriter<'a, T> {
    async fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Error> {
        self.1.run_with_timeout(Duration::from_secs(5), self.0.write(buf)).await.map_err(|_|
                embassy_net::tcp::Error::ConnectionReset).flatten()
    }
}
```
And wrapping the socket_tx:
```rust
  let socket_tx = WrappedWriter(socket_tx, EmbassyTimer);
```
... and I didn't have the 'permanent breaking' issue any more (just some occasional connection refused, as expected)

When I remove the wrapping I get the issue straight away.

**Comment by sammhicks** [comment]
I've implemented write timeouts in the [write_timeout](https://github.com/sammhicks/picoserve/tree/write_timeout) branch. Are you OK to test it before I publish the changes?

**Comment by flyaruu** [comment]
Yes, I can do that later today!

**Comment by flyaruu** [comment]
Yes, works! When dropping a connection (on a single socket) I get connection refused until the timeout elapses, then it accepts connections again.
